src/field_10x26_impl.h

```diff
@@ -1164,4 +1164,486 @@ static SECP256K1_INLINE void secp256k1_fe_from_storage(secp256k1_fe *r, const se
+        0xFF, 0x55, 0x33, 0x49, 0xC7, 0x5D, 0x3B, 0x11, 0x0F, 0xE5, 0xC3, 0x59,
+        0xD7, 0xED, 0xCB, 0x21, 0x1F, 0x75, 0x53, 0x69, 0xE7, 0x7D, 0x5B, 0x31,
+        0x2F, 0x05, 0xE3, 0x79, 0xF7, 0x0D, 0xEB, 0x41, 0x3F, 0x95, 0x73, 0x89,
+        0x07, 0x9D, 0x7B, 0x51, 0x4F, 0x25, 0x03, 0x99, 0x17, 0x2D, 0x0B, 0x61,
+        0x5F, 0xB5, 0x93, 0xA9, 0x27, 0xBD, 0x9B, 0x71, 0x6F, 0x45, 0x23, 0xB9,
+        0x37, 0x4D, 0x2B, 0x81, 0x7F, 0xD5, 0xB3, 0xC9, 0x47, 0xDD, 0xBB, 0x91,
+        0x8F, 0x65, 0x43, 0xD9, 0x57, 0x6D, 0x4B, 0xA1, 0x9F, 0xF5, 0xD3, 0xE9,
+        0x67, 0xFD, 0xDB, 0xB1, 0xAF, 0x85, 0x63, 0xF9, 0x77, 0x8D, 0x6B, 0xC1,
+        0xBF, 0x15, 0xF3, 0x09, 0x87, 0x1D, 0xFB, 0xD1, 0xCF, 0xA5, 0x83, 0x19,
+        0x97, 0xAD, 0x8B, 0xE1, 0xDF, 0x35, 0x13, 0x29, 0xA7, 0x3D, 0x1B, 0xF1,
+        0xEF, 0xC5, 0xA3, 0x39, 0xB7, 0xCD, 0xAB, 0x01
+    };
+
+    uint32_t u = 1, v = 0, q = 0, r = 1;
+    uint32_t f = f0, g = g0, m, w, x, y, z;
+    int i = 30, limit, zeros;
+
+    for (;;) {
+
+        /* Use a sentinel bit to count zeros only up to i. */
+        x = g | (UINT32_MAX << i);
+
+#if 0
+        zeros = __builtin_ctzl(x);
+#else
+        zeros = debruijn[((x & -x) * 0x04D7651F) >> 27];
+#endif
+
+        g >>= zeros;
+        u <<= zeros;
+        v <<= zeros;
+        eta -= zeros;
+        i -= zeros;
+
+        if (i <= 0) {
+            break;
+        }
+
+        VERIFY_CHECK((f & 1) == 1);
+        VERIFY_CHECK((g & 1) == 1);
+        VERIFY_CHECK((u * f0 + v * g0) == f << (30 - i));
+        VERIFY_CHECK((q * f0 + r * g0) == g << (30 - i));
+
+        if ((int32_t)eta < 0) {
+            eta = -eta;
+            x = f; f = g; g = -x;
+            y = u; u = q; q = -y;
+            z = v; v = r; r = -z;
+        }
+
+        /* Handle up to 8 divsteps at once, subject to eta and i. */
+        limit = ((int)eta + 1) > i ? i : ((int)eta + 1);
+        m = (UINT32_MAX >> (32 - limit)) & 255U;
+
+        w = (g * inv256[(f >> 1) & 127]) & m;
+
+        g += f * w;
+        q += u * w;
+        r += v * w;
+
+        VERIFY_CHECK((g & m) == 0);
+    }
+
+    t[0] = (int32_t)u;
+    t[1] = (int32_t)v;
+    t[2] = (int32_t)q;
+    t[3] = (int32_t)r;
+
+    return eta;
+}
+
+static void secp256k1_fe_update_de_30(int32_t *d, int32_t *e, const int32_t *t) {
+
+    /* P == 2^256 - 2^32 - C30 */
+    const int64_t C30 = 0x3D1L;
+    /* I30 == -P^-1 mod 2^30 */
+    const int32_t I30 = 0x12253531L;
+    const int32_t M30 = (int32_t)(UINT32_MAX >> 2);
+    const int32_t u = t[0], v = t[1], q = t[2], r = t[3];
+    int32_t di, ei, md, me;
+    int64_t cd, ce;
+    int i;
+
+    di = d[0];
+    ei = e[0];
+
+    cd = (int64_t)u * di + (int64_t)v * ei;
+    ce = (int64_t)q * di + (int64_t)r * ei;
+
+    /* Calculate the multiples of P to add, to zero the 30 bottom bits. We choose md, me
+     * from the centred range [-2^29, 2^29) to keep d, e within [-2^256, 2^256). */
+    md = (I30 * 4 * (int32_t)cd) >> 2;
```

Comment:
Is there a point to multiplying by 4 and then shifting down again?

Reply:
See the comment for this line: shift up/down is sign extension, so that the result is in [-2^29, 2^29].

Reply:
Ah, I see!